xarray/core/groupby.py

```diff
@@ -478,35 +480,78 @@ def _infer_concat_args(self, applied_example):
         return coord, dim, positions
 
     def _binary_op(self, other, f, reflexive=False):
+        from .dataarray import DataArray
+        from .dataset import Dataset
+
         g = f if not reflexive else lambda x, y: f(y, x)
-        applied = self._yield_binary_applied(g, other)
-        return self._combine(applied)
 
-    def _yield_binary_applied(self, func, other):
-        dummy = None
+        obj = self._obj
+        name = self._group.name
+        dim = self._group_dim
 
-        for group_value, obj in self:
-            try:
-                other_sel = other.sel(**{self._group.name: group_value})
-            except AttributeError:
-                raise TypeError(
-                    "GroupBy objects only support binary ops "
-                    "when the other argument is a Dataset or "
-                    "DataArray"
-                )
-            except (KeyError, ValueError):
-                if self._group.name not in other.dims:
-                    raise ValueError(
-                        "incompatible dimensions for a grouped "
-                        f"binary operation: the group variable {self._group.name!r} "
-                        "is not a dimension on the other argument"
+        if not isinstance(other, (Dataset, DataArray)):
+            raise TypeError(
+                "GroupBy objects only support binary ops "
+                "when the other argument is a Dataset or "
+                "DataArray"
+            )
+
+        if name not in other.dims:
+            raise ValueError(
+                "incompatible dimensions for a grouped "
+                f"binary operation: the group variable {name!r} "
+                "is not a dimension on the other argument"
+            )
+
+        group = self._group
+        if isinstance(group, _DummyGroup):
+            group = obj[dim]
```

Comment:
We might redefine group here but `name` for example has been defined by the earlier `group`, can this become an issue?

Reply:
No they should be the same, but I've reordered the  code to make it clearer. Thanks.